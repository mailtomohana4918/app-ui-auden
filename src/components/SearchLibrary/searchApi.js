import axios from "axios";
//on page load we need to load first 10 latest trending repos. github api q param is mandatory
// Hence passing javascript and it gives trending repos 
export const searchApi = {
  getTrendingRepos: async () => {
    try {
      const response = await axios.get(
        "https://api.github.com/search/repositories?q=javascript&sort=stars&order=desc&per_page=10"
      );

      return response.data.items;
    } catch (error) {
      console.log(error);
      throw error;
    }
  },
  //since we have a requirment user can search by username or repo, we don't have single api to achieve
  //the capabiility, Hence clubbed two api calls.
  getReposByUserOrRepoName: async (searchPattern) => {
    try {
      const userRepos = await axios.get(
        `https://api.github.com/users/${searchPattern}/repos`,
        { validateStatus: false }
      );

      const repoList = await axios.get(
        `https://api.github.com/search/repositories?q=${searchPattern}`,
        { validateStatus: false }
      );

      if (userRepos.status === 404) return repoList.data.items;
      else return [...userRepos.data, ...repoList.data.items];
    } catch (error) {
      console.log(error);
      throw error;
    }
  },
};
