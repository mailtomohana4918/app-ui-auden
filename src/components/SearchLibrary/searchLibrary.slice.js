import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  searchData: [],
};

export const searchSlice = createSlice({
  name: "search",
  initialState,
  reducers: {
    pushReposToStore: (state, action) => {
      state.searchData.push(...action.payload);
    },
  },
});

export const { pushReposToStore } = searchSlice.actions;
