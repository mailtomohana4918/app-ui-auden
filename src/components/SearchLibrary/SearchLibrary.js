import React, { useState, useEffect } from "react";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/core/styles";
import { Box, Button, Grid } from "@material-ui/core";
import Header from "../shared/Header/Header";
import { useSelector, useDispatch } from "react-redux";
import { pushReposToStore } from "./searchLibrary.slice";
import { searchApi } from "./searchApi";
import CustomTable from "../shared/CustomTable/CustomTable";

const columns = [
  { id: "name", label: "Name", minWidth: 100 },
  {
    id: "watchers",
    label: "Stars",
    minWidth: 170,
    align: "right",
    format: (value) => value.toLocaleString("en-US"),
  },
  {
    id: "forks",
    label: "Forks",
    minWidth: 170,
    align: "right",
    format: (value) => value,
  },
  {
    id: "open_issues",
    label: "Issues",
    minWidth: 170,
    align: "right",
    format: (value) => value,
  },
];

const useStyles = makeStyles((theme) => ({
  button: { paddingTop: "20px", paddingBottom: "20px" },
  boxContainer:{paddingTop:"120px" ,width:"20%"}
}));

const SearchLibrary = (props) => {
  const classes = useStyles();
  const [searchPattern, setSearchPattern] = useState("");
  const [isDisableButton, setIsDisableButton] = useState(true);

  const dispatch = useDispatch();
  const handleSearchPattern = (event) => {
    setSearchPattern(event.target.value);
    setIsDisableButton(false);
  };

  const searchData = useSelector((state) => state.search.searchData);
  const [filteredSearchData, setFilteredSearchData] = useState([]);
  const fetchTrendingRepos = async () => {
    const response = await searchApi.getTrendingRepos();
    dispatch(pushReposToStore(response));
  };
  useEffect(
    () => {
      if (searchData.length === 0) fetchTrendingRepos();
    },
    []
  );

  const fetchRepos = async () => {
    const filteredData = searchData.filter(
      (el) =>
        el.owner.login.toLowerCase().includes(searchPattern.toLowerCase()) ||
        el.name.toLowerCase().includes(searchPattern.toLowerCase())
    );
    // if found then return from store
    if (filteredData.length === 0) {
      //fetch from api
      const response = await searchApi.getReposByUserOrRepoName(searchPattern);

      dispatch(pushReposToStore(response));

      setFilteredSearchData(response);
    } else setFilteredSearchData(filteredData);

    //  setSearchPattern("")
  };

  return (
    <>
      <Header />
      <Grid
        container
        spacing={0}
        direction="column"
        alignItems="center"
        justify="center"
      >
        <Box className={classes.boxContainer}>
          <TextField
            placeholder="Search by keyword or username"
            variant="outlined"
            value={searchPattern}
            onChange={handleSearchPattern}
            fullWidth
            type="text"
            inputProps={{ "data-testid": "search-pattern-text-test-id" }}
          />
        </Box>
        <Box className={classes.button}>
          <Button
            variant="contained"
            color="primary"
            onClick={fetchRepos}
            disabled={isDisableButton}
            data-testid="search-button-test-id"
          >
            Search
          </Button>
        </Box>
        <Box>
          {
            <CustomTable
              searchResult={
                filteredSearchData.length === 0 && searchPattern === ""
                  ? searchData.slice(0, 10)
                  : filteredSearchData
              }
              columns={columns}
              source={"repo"}
            />
          }
        </Box>
      </Grid>
    </>
  );
};

export default SearchLibrary;
