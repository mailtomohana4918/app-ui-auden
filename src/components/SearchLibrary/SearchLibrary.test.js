import { render, fireEvent } from "@testing-library/react";
import SearchLibrary from "./SearchLibrary";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { BrowserRouter as Router } from "react-router-dom";
//import userEvent from '@testing-library/user-event'
describe("SearchLibrary", () => {
  const initialState = {
    search: {
      searchData: [
        { id: 1, name: "React", watchers: 2000, forks: 100, open_issues: 200 },

        { id: 2, name: "Vue", watchers: 3000, forks: 200, open_issues: 300 },
        {
          id: 3,
          name: "Angular",
          watchers: 4000,
          forks: 300,
          open_issues: 400,
        },
        { id: 4, name: "Django", watchers: 5000, forks: 400, open_issues: 500 },
      ],
    },
  };
  const mockStore = configureStore();
  let store;
  it("render component as expected", () => {
    store = mockStore(initialState);
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <SearchLibrary />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe("Search Button", () => {
    it("search button should be disabled on page load", () => {
      const { getByTestId } = render(
        <Provider store={store}>
          <Router>
            <SearchLibrary />
          </Router>
        </Provider>
      );

      expect(getByTestId("search-button-test-id").disabled).toBeTruthy();
    });

    it("search button should be enabled upon search pattern provided", () => {
      const { getByTestId } = render(
        <Provider store={store}>
          <Router>
            <SearchLibrary />
          </Router>
        </Provider>
      );

      const searchPatternInput = getByTestId("search-pattern-text-test-id");

      fireEvent.change(searchPatternInput, { target: { value: "React" } });

      expect(getByTestId("search-button-test-id")).toBeEnabled();
    });
  });
});
