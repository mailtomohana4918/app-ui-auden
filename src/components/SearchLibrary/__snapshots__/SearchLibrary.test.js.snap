// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchLibrary render component as expected 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-colorInherit"
        >
          Github Library
        </h4>
      </div>
    </header>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiBox-root MuiBox-root-3 makeStyles-boxContainer-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="search-pattern-text-test-id"
            placeholder="Search by keyword or username"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-5"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-8 makeStyles-button-1"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="search-button-test-id"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Search
        </span>
      </button>
    </div>
    <div
      class="MuiBox-root MuiBox-root-9"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
      >
        <h5
          class="MuiTypography-root makeStyles-typoGraphy-11 MuiTypography-h5 MuiTypography-colorPrimary"
        >
           Repository Details
        </h5>
        <div
          class="MuiPaper-root makeStyles-root-10 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiTableContainer-root"
          >
            <table
              aria-label="sticky table"
              class="MuiTable-root makeStyles-table-12 MuiTable-stickyHeader"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
                    scope="col"
                    style="min-width: 100px;"
                  >
                    Name
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
                    scope="col"
                    style="min-width: 170px;"
                  >
                    Stars
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
                    scope="col"
                    style="min-width: 170px;"
                  >
                    Forks
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-stickyHeader"
                    scope="col"
                    style="min-width: 170px;"
                  >
                    Issues
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="checkbox"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    React
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2,000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    100
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    200
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="checkbox"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    Vue
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    3,000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    200
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    300
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="checkbox"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    Angular
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    4,000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    300
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    400
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="checkbox"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    Django
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    5,000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    400
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    500
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="MuiTablePagination-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
            >
              <div
                class="MuiTablePagination-spacer"
              />
              <p
                class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
                id="mui-5145"
              >
                Rows per page:
              </p>
              <div
                class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
              >
                <div
                  aria-haspopup="listbox"
                  aria-labelledby="mui-5145 mui-91655"
                  class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
                  id="mui-91655"
                  role="button"
                  tabindex="0"
                >
                  10
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput"
                  tabindex="-1"
                  value="10"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
              >
                1-4 of 4
              </p>
              <div
                class="MuiTablePagination-actions"
              >
                <button
                  aria-label="Previous page"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
                  disabled=""
                  tabindex="-1"
                  title="Previous page"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                      />
                    </svg>
                  </span>
                </button>
                <button
                  aria-label="Next page"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
                  disabled=""
                  tabindex="-1"
                  title="Next page"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
